AbletonMCP_Remote_Script/handlers/midi.py

```
@@ -0,0 +1,446 @@
+        if ctrl:
+            ctrl.log_message("Error capturing MIDI: " + str(e))
+        raise
+
+
+def apply_groove(song, track_index, clip_index, groove_amount, ctrl=None):
+    """Apply groove amount (sets global song.groove_amount, validates clip exists)."""
+    try:
+        _get_midi_clip(song, track_index, clip_index)  # validate clip exists
+        groove_amount = float(groove_amount)
+        if groove_amount < 0.0 or groove_amount > 1.0:
+            raise ValueError("groove_amount must be 0.0-1.0, got {0}".format(groove_amount))
```

Ruff (0.15.0)

[warning] Abstract `raise` to an inner function

(TRY301)